### PR TITLE
Adding mechanism to query path structure and active values

### LINKED ISSUE #123: Add support for retrieving settings at run-time
Add a mechanism to retrieve:
1. What paths are available at run-time
2. What are the values associated with those paths

During the meeting today, we discussed opening up command endpoints to the Miniconf MQTT client. We could expose two endpoints:
1. Request a list of settings paths be published
2. Request the value of a specific settings path

CC @jordens @nkrackow

### PR DESCRIPTION
**Pull request by ryan-summers**
This PR fixes #123 by adding two command: `<prefix>/list` and a publication to the path tree.

When a response topic is provided, `list` publishes all of the settings paths sequentially to the provided topic. To indicate completion, an empty message is sent at the end.

When a query is requested, the payload is expected to be a Miniconf path. In this case, the settings structure queries the provided path and the value of the path is published to the provided response topic.

### CODE REVIEW

File: src/mqtt_client/mod.rs [hunk 1]
```diff
@@ -339,58 +386,96 @@ where
...
+
+            // For paths, we do not want to include the leading slash.
+            let path = path.strip_prefix('/').unwrap_or(path);
+
+            let command = {
+                match path.split_once('/') {
+                    Some((root, tail)) => (root, Some(tail)),
+                    None => (path, None),
+                }
+            };
```
Comment (ryan-summers) [inline src/mqtt_client/mod.rs:407]: Transform this into an enum that implements something like `FromStr`

File: src/mqtt_client/mod.rs [hunk 2]
```diff
@@ -55,15 +55,15 @@ mod sm {
...
+        pub iteration_state: super::MiniconfIter<M>,
     }
 
     impl<C: embedded_time::Clock, M: super::Miniconf> Context<C, M> {
         pub fn new(clock: C) -> Self {
             Self {
                 clock,
                 timeout: None,
-                republish_state: Default::default(),
+                iteration_state: Default::default(),
```
Comment (ryan-summers) [inline src/mqtt_client/mod.rs:66]: Revert these changes, not needed.

File: src/mqtt_client/mod.rs [hunk 3]
```diff
@@ -89,6 +89,31 @@ mod sm {
     }
 }
 
+enum Command<'a> {
+    Query,
+    List,
+    Modify(&'a str),
+}
```
Comment (ryan-summers) [inline src/mqtt_client/mod.rs:96]: It seems like these commands need to be listed via some retained settings so that the client can know what commands are available. This would aid us in the future if we expose new APIs over MQTT.

We could do this via the `alive` topic, or we could just publish the `miniconf` version being used on the `alive` topic.

Any thoughts on this one @jordens?

Comment (jordens) [reply src/mqtt_client/mod.rs:96]: The latter. I'd make `alive` contain the `miniconf` protocol version (thus guaranteeing the availability of the commands) and (probably) also include an application version identifier.

Comment (jordens) [reply src/mqtt_client/mod.rs:96]: But this change to alive can also be done in a separate PR.

File: src/mqtt_client/mod.rs [hunk 4]
```diff
@@ -339,58 +409,82 @@ where
...
+            };
+
+            let Ok(command) = Command::from_str(path) else {
+                info!("Unknown Miniconf command: {path}");
+                return;
+            };
+
+            let mut data = [0u8; MESSAGE_SIZE];
+            let message = match command {
+                Command::Query => {
```
Comment (jordens) [inline src/mqtt_client/mod.rs:429]: Can we stick to `Get` as in the miniconf trait?

File: src/mqtt_client/mod.rs [hunk 5]
```diff
@@ -339,58 +409,82 @@ where
...
+                        .get(path, &mut data)
+                        .map(|len| &data[..len])
+                        .unwrap_or(b"Invalid path")
                 }
-                None => {
-                    info!("Unexpected MQTT topic: {}", topic);
+
+                Command::List => {
+                    let Some(response_topic) = properties.into_iter().find_map(|p| {
+                        if let Ok(minimq::Property::ResponseTopic(topic)) = p {
```
Comment (jordens) [inline src/mqtt_client/mod.rs:441]: Also carry over correlation data into the response(s)
I think we already do this for the set response and/or in minireq.

Comment (ryan-summers) [reply src/mqtt_client/mod.rs:441]: This is an interesting problem, as we will have to cache the `CorrelationData` outside of the minimq handler. We should already have enough buffer space for this (because we reserve the max MQTT message size), but I'll have to look at how we can cache both the response topic and CD. It's definitely possible, but may require minimq updates

File: src/mqtt_client/mod.rs (hunk 4 above)
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:429]: Sure. I was hopping back and forth between the naming, but `get` is definitely simpler. In reality, I imagine tooling will abstract all this so the user won't even notice it.

File: src/mqtt_client/mod.rs (hunk 5 above)
Comment (jordens) [reply src/mqtt_client/mod.rs:441]: Can't you just save and then copy/clone the empty-payload `Publication` and take things from there?

File: src/mqtt_client/mod.rs (hunk 3 above)
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:96]: Let's follow up with this in a separate PR. I'll open an issue.

Comment (ryan-summers) [reply src/mqtt_client/mod.rs:96]: Added in https://github.com/quartiq/miniconf/issues/125

File: src/mqtt_client/mod.rs [hunk 6]
```diff
@@ -339,58 +415,89 @@ where
...
-                        msg = String::from("Configuration Error");
+
+                Command::List => {
+                    if !properties
+                        .into_iter()
+                        .any(|prop| matches!(prop, Ok(minimq::Property::ResponseTopic(_))))
+                    {
+                        // If there's no response topic, there's no where we can publish the list.
+                        // Ignore the request.
+                        return;
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:452]: The fundamental issue here is that if the client provides no `ResponseTopic`, it's not clear where the response status should be published. Previously, we used a default `log` topic, which has since been removed.

Do we want to bring back a default topic?

[the remark this answers is not in the document]
Comment (jordens) [reply src/mqtt_client/mod.rs:452]: Ah no. Ignore and return here is fine.
My beef was with the return in L466 and https://github.com/quartiq/miniconf/pull/124/files#r1044566863

File: py/miniconf-mqtt/miniconf/miniconf.py [hunk 7]
```diff
@@ -39,8 +39,11 @@ def __init__(self, client, prefix):
         self.prefix = prefix
         self.inflight = {}
         self.client.on_message = self._handle_response
-        self.response_topic = f'{prefix}/response'
-        self.client.subscribe(self.response_topic)
+        self.command_response_topic = f'{prefix}/response/command'
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply py/miniconf-mqtt/miniconf/miniconf.py:42]: The fine grained topics are because different topics require different behavior upon value reception. For example, a `list` command isn't complete after a single reception, whereas the `get` and `set` are

[the remark this answers is not in the document]
Comment (jordens) [reply py/miniconf-mqtt/miniconf/miniconf.py:43]: Right. But that could alo be done depending on what's in `inflight`, e.g. like https://github.com/quartiq/miniconf/pull/124#discussion_r1044487541

File: src/mqtt_client/mod.rs [hunk 8]
```diff
@@ -89,6 +89,36 @@ mod sm {
     }
 }
 
+enum Command<'a> {
+    List,
+    Get(&'a str),
```
Comment (jordens) [inline src/mqtt_client/mod.rs:94]: I'd `Get { path: &'a str }` for symmetry.

File: src/mqtt_client/mod.rs [hunk 9]
```diff
@@ -199,11 +229,56 @@ where
...
+            let response: Response<MAX_TOPIC_LENGTH> = iter
+                .next()
+                .map(|path| Response::custom(ResponseCode::Continue, &path))
+                .unwrap_or_else(Response::ok);
+
+            let message = match serde_json_core::to_slice(&response, &mut data) {
+                Ok(len) => &data[..len],
+
+                // If the message buffer is too small, we can't ever provide a response.
+                _ => return,
```
Comment (jordens) [inline src/mqtt_client/mod.rs:258]: Not error here?

**Review by jordens — COMMENTED** [review]
Just some comments. Looking really good.

File: py/miniconf-mqtt/miniconf/miniconf.py [hunk 10]
```diff
@@ -98,3 +139,46 @@ async def command(self, path, value, retain=False, timeout=5):
...
+        assert request_id not in self.inflight
+        self.inflight[request_id] = ([], fut)
+
+        self.client.publish(f'{self.prefix}/list', payload='',
+                            correlation_data=request_id,
+                            response_topic=self.response_topic)
+        return fut
+
+
+    async def get(self, path, timeout=5):
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply py/miniconf-mqtt/miniconf/miniconf.py:159]: Do we want to similarly remove the timeout for `set` then? I'm confused why we wouldn't want a timeout in these.

File: src/mqtt_client/mod.rs [hunk 11]
```diff
@@ -525,39 +548,80 @@ where
...
+                code: code as u8,
+                msg,
+            },
+            Err(_) => Self::error("Message too long"),
+        }
+    }
+
+    pub fn error(message: &str) -> Self {
+        let mut msg = String::new();
+        msg.push_str(message).ok();
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:587]: That's the intent. If we can't serialize the message (because the user specified some very small max topic size), there's nothing more we can necessarily do and we fall back on just providing the response code. I'll document this.

File: py/miniconf-mqtt/miniconf/miniconf.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jordens) [reply py/miniconf-mqtt/miniconf/miniconf.py:159]: Yes. Timeouts should be higher level things.
Because the user can and should wrap this code in a timeouter themselves. https://github.com/quartiq/miniconf/pull/124#discussion_r1044348759

File: src/mqtt_client/mod.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jordens) [reply src/mqtt_client/mod.rs:587]: But shouldn't we do it like L602?

File: src/mqtt_client/mod.rs (hunk 9 above)
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:258]: The `poll()` closure currently returns `()`, so there's no chance to return an error. We could refactor it, but I'd rather keep that into a separate PR, since there's other rework around the `poll()` function we could complete as well. I think this is maintaining current behavior.

Comment (jordens) [reply src/mqtt_client/mod.rs:258]: I meant still providing a fixed error response to the user. But I'm unsure whether it's better to abort a listing when we hit an error or silently discard paths that are too long.

Comment (ryan-summers) [reply src/mqtt_client/mod.rs:265]: If we can't enqueue a response of 32-bytes, it doesn't make sense to e.g. try something progressively smaller until it eventually works. I would just say that we have a static limit on the minimum transport size

File: py/miniconf-mqtt/miniconf/__main__.py [hunk 12]
```diff
@@ -61,11 +63,21 @@ def main():
...
 
+    async def list_paths():
+        interface = await Miniconf.create(prefix, args.broker)
+        for path in await interface.list_paths():
+            value = await interface.get(path)
+            print(f'{path} = {value}')
+
+    if args.list:
+        loop.run_until_complete(list_paths())
+        return
```
Comment (ryan-summers) [inline py/miniconf-mqtt/miniconf/__main__.py:74]: We could alternatively just place this as an activity to be completed immediately after all the `set()` calls complete (i.e. if none are supplied, just list settings), which would allow the user to view settings after application as well.

File: src/mqtt_client/mod.rs (hunk 9 above)
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:265]: Ah hmm, this is for the list command, I was thinking this was in generic `get/set/list` responses.

Indeed, we likely need to generate an error if we could not serialize the response into the buffer, since CD or response topic may be taking too much space.

File: py/miniconf-mqtt/miniconf/__main__.py (hunk 12 above)
Comment (jordens) [reply py/miniconf-mqtt/miniconf/__main__.py]: I would be a bit hesitant to default to getting all paths as well. What about  behaving similar to e.g. Linux tools like sysctl:

```
rj@grouch:~/src/sinara-systems (master)$ sysctl kernel.pty
kernel.pty.max = 4096
kernel.pty.nr = 15
kernel.pty.reserve = 1024
rj@grouch:~/src/sinara-systems (master)$ sysctl kernel.pty.max
kernel.pty.max = 4096
rj@grouch:~/src/sinara-systems (master)$ sysctl kernel.pty.max=9
sysctl: permission denied on key "kernel.pty.max", ignoring
rj@grouch:~/src/sinara-systems (master)$ sysctl -N kernel.pty.max
kernel.pty.max
rj@grouch:~/src/sinara-systems (master)$ sysctl -h

Usage:
 sysctl [options] [variable[=value] ...]

Options:
  -a, --all            display all variables
  -A                   alias of -a
  -X                   alias of -a
      --deprecated     include deprecated parameters to listing
  -b, --binary         print value without new line
  -e, --ignore         ignore unknown variables errors
  -N, --names          print variable names without values
  -n, --values         print only values of the given variable(s)
  -p, --load[=<file>]  read values from file
  -f                   alias of -p
      --system         read values from all system directories
  -r, --pattern <expression>
                       select setting that match expression
  -q, --quiet          do not echo variable set
  -w, --write          enable writing a value to variable
  -o                   does nothing
  -x                   does nothing
  -d                   alias of -h

 -h, --help     display this help and exit
 -V, --version  output version information and exit

For more details see sysctl(8).
```

File: src/mqtt_client/mod.rs [hunk 13]
```diff
@@ -141,8 +170,10 @@ where
     mqtt: minimq::Minimq<Stack, Clock, MESSAGE_SIZE, 1>,
     settings: Settings,
     state: sm::StateMachine<sm::Context<Clock, Settings>>,
-    settings_prefix: String<MAX_TOPIC_LENGTH>,
     prefix: String<MAX_TOPIC_LENGTH>,
+    listing_state: Option<MiniconfIter<Settings>>,
+    properties_cache: Option<Vec<u8, MESSAGE_SIZE>>,
```
Comment (jordens) [inline src/mqtt_client/mod.rs:175]: I suspect a single Option for both both doesn't work, right?

```
pending_response: Option<(Vec<u8, MESSAGE_SIZE>, Response<32>)>,
```

File: src/mqtt_client/mod.rs [hunk 14]
```diff
@@ -302,14 +388,12 @@ where
         if !self.mqtt.client().is_connected() {
```
Comment (jordens) [inline src/mqtt_client/mod.rs:388]: Then invert the `if` arms.
But doesn't this cause a lot of traffic through the `process_event(sm::Events::Connected)`. Maybe that's fine though.

File: src/mqtt_client/mod.rs [hunk 15]
```diff
@@ -428,39 +624,103 @@ where
...
 
+#[repr(u8)]
+pub enum ResponseCode {
+    Ok = 0,
+    Continue = 1,
+
+    Error = 0xFF,
+}
+
+impl From<ResponseCode> for u8 {
```
Comment (jordens) [inline src/mqtt_client/mod.rs:635]: Doesn't `code as u8` work?

**Review by jordens — CHANGES_REQUESTED** [review]
A couple nits. But do feel free to defer and discuss them later.

File: py/miniconf-mqtt/miniconf/miniconf.py [hunk 16]
```diff
@@ -39,8 +43,27 @@ def __init__(self, client, prefix):
...
+        self.client.on_subscribe = self._handle_subscription
         self.response_topic = f'{prefix}/response'
-        self.client.subscribe(self.response_topic)
+        mid = self.client.subscribe(f'{prefix}/response')
+        self._pending_subscriptions = {mid: asyncio.get_running_loop().create_future()}
+
+
+    async def subscriptions_complete(self):
+        """ Wait for all pending subscriptions to complete. """
+        for subscription in list(self._pending_subscriptions.values()):
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply py/miniconf-mqtt/miniconf/miniconf.py:54]: This listification is required because we're modifying `self._pending_subscriptions` when we do `await`. The `list()` call forces a complete iteration and copy of the values in the dict before the first `await`

File: src/mqtt_client/mod.rs (hunk 13 above)
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:175]: The properties cache has to be used for both listing and the pending response, so it belongs to more than just one cached thing, hence why it got broken out.

File: src/mqtt_client/mod.rs [hunk 17]
```diff
@@ -428,39 +624,103 @@ where
...
+    fn from(code: ResponseCode) -> u8 {
+        code as u8
+    }
+}
+
 /// The payload of the MQTT response message to a settings update request.
 #[derive(Serialize)]
-pub struct SettingsResponse {
+struct Response<const N: usize> {
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:643]: Oooh that's a really nice idea. I'll open an issue for this since it would just bloat this PR further

File: py/miniconf-mqtt/miniconf/miniconf.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jordens) [reply py/miniconf-mqtt/miniconf/miniconf.py:54]: Ah. Ack!

File: src/mqtt_client/mod.rs (hunk 13 above)
Comment (jordens) [reply src/mqtt_client/mod.rs:175]: Ack

File: src/mqtt_client/mod.rs (hunk 15 above)
Comment (jordens) [reply src/mqtt_client/mod.rs:635]: Sounds good.

File: src/mqtt_client/mod.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:643]: Added https://github.com/quartiq/miniconf/issues/131

File: src/mqtt_client/mod.rs [hunk 18]
```diff
@@ -428,39 +624,103 @@ where
...
-    pub fn error(msg: String<64>) -> Self {
-        Self { code: 255, msg }
+    /// Generate a custom response with any response code.
+    ///
+    /// # Args
+    /// * `code` - The code to provide in the response.
+    /// * `msg` - The message to provide in the response.
+    pub fn custom(code: impl Into<u8>, message: &str) -> Self {
+        // Truncate the provided message to ensure it fits within the heapless String.
+        let message = if message.len() > N {
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:663]: It's a possibility, but it then precludes some use cases of `Response` in `handled_update()` by user software (or atleast makes it harder)

File: py/miniconf-mqtt/miniconf/miniconf.py [hunk 19]
```diff
@@ -52,28 +75,57 @@ def _handle_response(self, _client, topic, payload, _qos, properties):
...
-                return
+        response = json.loads(payload)
 
-            self.inflight[request_id].set_result(json.loads(payload))
-            del self.inflight[request_id]
+        if topic == self.response_topic:
+            # Handle list responses
+            if isinstance(self.inflight[request_id], tuple):
+                if response['code'] == 0xFF:
+                    handler[1].set_result(response)
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply py/miniconf-mqtt/miniconf/miniconf.py:97]: Oh, that's perfect. Didn't know that was available. Will update.

File: py/miniconf-mqtt/miniconf/miniconf.py [hunk 20]
```diff
@@ -95,6 +147,48 @@ async def command(self, path, value, retain=False, timeout=5):
...
+        await self._pending_subscriptions[subscription]
+
+        self.client.publish(
+            f'{self.prefix}/settings/{path}', payload='', qos=0,
+            response_topic=self.response_topic,
+            correlation_data=request_id)
+
+        response, value = await fut
+
+        self.client.unsubscribe(f'{self.prefix}/settings/{path}')
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply py/miniconf-mqtt/miniconf/miniconf.py:189]: There's an ack for it, but I figured it wasn't all that important. In reality, the ACK will come relatively quick, and the only error case would be calling `get()` multiple times with the same path, in which case the first `subscribe` will await, and we'll process the unsuback.

I don't think there's really any harm in exiting before it's ACKed

File: src/mqtt_client/mod.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (jordens) [reply src/mqtt_client/mod.rs:643]: Ack. But I would leave the topics we publish to as it is now (`list()` payloads to response topic and `get()` payload to respective settings topics), right?

File: py/miniconf-mqtt/miniconf/miniconf.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jordens) [reply py/miniconf-mqtt/miniconf/miniconf.py:189]: But with asyncio I can imagine the following message sequence being sent (two concurrent get on the same topic): "sub1,pub1,time passes, sub2,suback2,resp1,unsub1,unsuback1,time passes pub2", i.e. an overlap due to asyncio and "pub2" not being seen as "unsub1" went in after "sub2".

File: src/mqtt_client/mod.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:643]: Yup, agreed

File: py/miniconf-mqtt/miniconf/miniconf.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jordens) [reply py/miniconf-mqtt/miniconf/miniconf.py:189]: In fact this may be a big problem and waiting for unsuback.would.not help.
What about subscribing to all settings on creation and then never unsubscribing?

[the remark this answers is not in the document]
Comment (ryan-summers) [reply py/miniconf-mqtt/miniconf/miniconf.py:189]: In your scenario, it's important to note that `sub1` and `sub2` are two unique subscriptions. unsub1 does not result in cancelling out `sub2` even if they're on the same topics because they use unique identifiers to denote which subscription is being canceled.

File: src/mqtt_client/mod.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:663]: > I don't think you need to check for that. Just indiscriminately truncate with `String::from(&message[..N])` Same in error().

This will panic if N is larger than `message.len()`, hence the bounds check first.

See https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=aadc07367bb233832d951ef314ded976

File: src/mqtt_client/mod.rs [hunk 21]
```diff
@@ -199,11 +230,69 @@ where
...
+                .map(|path| Response::custom(ResponseCode::Continue, &path))
+                .unwrap_or_else(Response::ok);
+
+            let message = match serde_json_core::to_slice(&response, &mut data) {
+                Ok(len) => &data[..len],
+
+                _ => {
+                    let response: Response<32> = Response::error("Buffer too small");
+                    let Ok(len) = serde_json_core::to_slice(&response, &mut data) else {
+                        log::warn!("Message size is too small for responses");
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:267]: Honestly, I don't think we need the level of detail here. We'll get a line number and unique message for users to refer back to. They won't have any clue what the "buffer too small" response is and I would think it would just obfuscate what they need to do (i.e. increase the message size)

File: src/mqtt_client/mod.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (ryan-summers) [reply src/mqtt_client/mod.rs:663]: Also, let's hold off on using a `&str`. Using the string slice prevents us from e.g. sending in a debug-formatted `miniconf::Error`, which means users may not see why a serialization/deserialization op is failing

File: py/miniconf-mqtt/miniconf/miniconf.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jordens) [reply py/miniconf-mqtt/miniconf/miniconf.py:189]: How does `self.client.unsubscribe(f'{self.prefix}/settings/{path}')` know which one to unsubscribe?

File: src/mqtt_client/mod.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (jordens) [reply src/mqtt_client/mod.rs:663]: > This will panic if N is larger than message.len(), hence the bounds check first.

Oh. My fault. I was thinking Python!

File: py/miniconf-mqtt/miniconf/miniconf.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (ryan-summers) [reply py/miniconf-mqtt/miniconf/miniconf.py:189]: Oh huh, that's a really good point. It doesn't look like GMQTT distinguishes. I'll get this fixed up

[the remark this answers is not in the document]
Comment (jordens) [reply py/miniconf-mqtt/miniconf/miniconf.py]: I would still do a wildcard subscribe to all settings and remain subscribed. Seems easier to me an fine w.r.t. traffic.

[the remark this answers is not in the document]
Comment (ryan-summers) [reply py/miniconf-mqtt/miniconf/miniconf.py]: I've updated to just subscribe to the settings/# path and ignore local traffic. It will get undesired traffic from e.g. other miniconf host-side clients, but it's not problematic (log and ignore). The unsubscribe + unsub ack actually added a lot of latency during the settings tree retrieval because so many more messages were being transferred.

**Review by jordens — APPROVED** [review]
Nice. If you want, maybe do https://github.com/quartiq/miniconf/pull/124#discussion_r1048205692 but it's a minor nit.

File: src/mqtt_client/mod.rs [hunk 22]
```diff
@@ -662,15 +662,9 @@ impl<const N: usize> Response<N> {
...
-        let message = if message.len() > N {
-            &message[..N]
-        } else {
-            message
-        };
-
         Self {
             code: code.into(),
-            msg: String::from(message),
+            msg: String::from(&message[..N.min(message.len())]),
```
Comment (jordens) [inline src/mqtt_client/mod.rs:667]: Same in L677
